## Combinatorial Optimization

#### Assignment IV

***Heuristics for the Knapsack Problem - Solving Multiple Instances***

**Instance:** C500I500

Guilherme Cadori

09/07/2023


In [1]:
# Importing required libraries
import numpy as np


#### 1. Read the instance

In [2]:
# Reading the target file
# GC: the code template below was provided in class and was adapted by GC (2023-07-09)

filename = "C:\gccadori\Lab4_CO\Knap_C500I500.dat"  # Substitua pelo nome do arquivo real

# Variáveis para armazenar os dados lidos
linha = None
coluna = None
cost = None
rhs = None
matrix = None

# Função auxiliar para converter uma linha de valores em uma lista de floats
def parse_values(line):
    return [float(value) for value in line.strip().split()]

# Ler o arquivo e extrair os dados
with open(filename, "r") as file:
    lines = file.readlines()

# Ler o escalar
linha = int(lines[0].strip())
coluna = int(lines[2].strip())

# Ler o vetor
cost = np.array(parse_values(lines[5]))

# Ler a matriz
end = 7 + linha
matrix_lines = [line.strip() for line in lines[7:end]]
matrix = np.array([parse_values(line) for line in matrix_lines],dtype=object)

# Ler vetor de RHS
rhs = np.array(parse_values(lines[end+1]))

# Imprimir os dados lidos
print("\n linhas:", linha)
print("\n colunas:", coluna)
print("\n cost:", cost)
print("\n rhs:", rhs)
print("\n Matrix:")
for row in matrix:
    print(row)



 linhas: 500

 colunas: 500

 cost: [0.30916 0.98209 0.94039 0.18065 0.16531 0.43426 0.53761 0.53339 0.75291
 0.12222 0.19591 0.57626 0.19449 0.8141  0.50288 0.86982 0.10779 0.56491
 0.45191 0.25363 0.80109 0.92301 0.0753  0.57091 0.22717 0.0089  0.52606
 0.55681 0.27917 0.9448  0.30023 0.9987  0.10447 0.86009 0.4587  0.41327
 0.87781 0.42382 0.11196 0.65734 0.93795 0.13805 0.19439 0.0731  0.65883
 0.95469 0.46864 0.42741 0.54299 0.08467 0.98709 0.07894 0.77121 0.76676
 0.01637 0.10572 0.89093 0.90917 0.41176 0.42202 0.82331 0.42036 0.91505
 0.2701  0.49529 0.38112 0.41869 0.90669 0.75153 0.88346 0.35695 0.24912
 0.94427 0.38372 0.18051 0.76134 0.89264 0.54002 0.18292 0.36996 0.96167
 0.83045 0.45323 0.47259 0.7401  0.85425 0.37873 0.29918 0.29172 0.0157
 0.95315 0.54283 0.32504 0.97377 0.1087  0.9828  0.84225 0.69351 0.89178
 0.11858 0.92069 0.08849 0.28848 0.49402 0.05352 0.58145 0.37681 0.02892
 0.08139 0.99204 0.24774 0.77752 0.82395 0.17793 0.55099 0.45361 0.86113
 0.93196 0.3890

[2.67966 11.97006 5.80113 24.60847 19.5494 16.75184 23.18167 14.36185
 4.62848 15.86543 0.28766 9.76192 18.50814 16.26115 1.14859 4.29799
 11.38786 20.74365 13.60985 15.70032 0.3422 1.35028 19.18164 10.81215
 19.80886 2.41442 4.07835 19.78946 1.47177 10.98844 7.72747 0.55634
 0.36293 24.70824 21.36212 8.18731 4.10087 23.36542 2.62916 13.29286
 13.11135 12.44935 11.28011 9.82707 13.61677 7.03598 3.65939 3.33245
 8.46638 19.3702 5.05597 0.64048 14.57709 22.16125 14.12545 6.38194
 11.3463 22.26603 0.09277 9.19198 14.61347 8.54313 9.36717 9.06224 9.08192
 14.87164 22.60219 24.93035 4.33873 21.03355 10.95086 1.05201 6.1184
 6.97231 8.64679 1.62908 4.98172 2.84878 4.40094 16.54081 1.39732 9.8222
 6.74551 21.8111 4.16417 12.2349 7.03417 23.26999 23.67798 5.78056 3.84045
 21.51146 18.06793 17.68456 24.40889 15.1464 15.60165 16.95451 4.4641
 3.20208 17.41541 0.75767 9.17742 19.86627 17.39009 0.28082 19.80546
 20.43766 20.82689 12.48335 7.64635 12.22336 12.95508 11.04583 22.33765
 3.94411 13.610

[6.64465 1.64554 6.61551 11.83516 13.47124 11.09758 16.95861 23.33185
 13.39196 3.6678 19.66102 17.84467 15.2784 9.13824 11.41367 4.5638 3.84085
 3.18265 15.87854 20.6612 2.78134 20.93957 6.41208 17.78537 18.74876
 10.45842 24.63991 22.92852 9.69297 9.7273 11.65711 20.97854 11.25026
 8.17265 7.76081 10.98405 8.96961 2.26798 18.0151 4.73621 1.40576 1.60158
 17.7004 15.58586 1.49404 10.40097 9.07001 14.71948 15.34288 17.75013
 1.46616 16.74643 7.29776 3.50859 18.93436 4.74222 2.42403 15.7367
 11.72623 7.7649 4.6534 9.67656 8.92546 10.13761 7.75398 21.20252 0.7687
 19.61267 5.18031 15.43232 20.97907 20.26017 12.76281 4.49786 20.48218
 18.9325 23.55726 1.9534 5.76988 24.41057 18.43071 14.87956 5.69879
 4.57462 10.67891 5.44082 18.82455 9.29256 5.09455 24.1487 17.20132
 2.58871 8.42438 13.57003 21.49586 5.90256 4.31868 9.00759 15.62227
 13.51299 12.81865 18.02354 21.57584 0.20161 13.53731 21.63629 16.05965
 14.51562 14.09401 2.99323 7.27138 10.10388 15.9511 15.20206 1.00363
 18.02849 4.759 

#### 2.	Construct an initial feasible solution

In [3]:
# Renaming variables for testing purposes
c = cost

A = matrix

b = rhs


In [4]:
# Providing an initial feasible solution

def randomStart(nVars, A, b, c):
    # Creating variable names x1, ..., x_nVars
    variable_names = ['x{}'.format(i) for i in range(1, nVars+1)]

    while True:
        np.random.seed(12345)
        x = np.random.randint(2, size=nVars)
        
        if np.all(A @ x <= b):
            FO = x @ c
            print('The initial feasible solution is:')   
            print('Objective function value:', round(FO, 2))
            return


In [5]:
# Testing the function
randomStart(nVars = coluna, A = A, b = b, c = c)


The initial feasible solution is:
Objective function value: 128.17


#### 3.	Program the heuristic you proposed in the previous assingment

In [6]:
# Implementing the heuristic algorithm proposed in Lab III

def naiveKnapsack(nVars, items_c, capacity_b, A):
    # Sort the items in descending order of value
    sorted_items = sorted(range(nVars), key=lambda i: items_c[i], reverse=True)

    # Initialize variables
    selected_items = np.zeros(nVars, dtype=int)
    total_weight = np.zeros_like(capacity_b, dtype=items_c.dtype)
    total_value = 0

    # Greedy selection process
    for idx in sorted_items:
        weight = A[:, idx]
        if np.all(total_weight + weight <= capacity_b):
            # Add the item to the knapsack
            selected_items[idx] = 1
            total_weight += weight.astype(total_weight.dtype)  # Ensure matching data type
            total_value += items_c[idx]

    # Print the selected items and total value
    print('The initial feasible solution is:')
    print('Objective function value:', round(total_value, 2))

    # Return the complete list of items
    return


In [7]:
# Testing the function
naiveKnapsack(nVars = coluna, items_c = c, capacity_b = b , A = A)


The initial feasible solution is:
Objective function value: 230.31


#### 4.	Program a Local Search 

In [8]:
# Local Search metaheuristic implementation
def LocalSearch(nVars, A, b, c, nStarts=1, max_iterations=1000):
    import numpy as np

    # Creating variable names
    variable_names = ['x{}'.format(i) for i in range(1, nVars + 1)]

    best_solution = None
    best_value = -np.inf

    # Escaping Local Optima - Multiple Starting Solutions
    for n in range(1, nStarts + 1):
        # Producing an initial feasible solution
        while True:
            x = np.random.randint(2, size=nVars)
            if np.all(A @ x <= b):
                FO = x @ c
                break  # Exit the loop once an initial feasible solution is found

        # Setting initial search parameters
        s = FO
        s_new = 0
        n_iterations = 0
        best_iteration = 0

        # Termination criteria
        while n_iterations - best_iteration < max_iterations:
            n_iterations += 1

            # Neighbor Selection: Swap Two - First Improvement Strategy
            def generate_neighbor_solution(x):
                x_new = x.copy()
                idx1, idx2 = np.random.choice(len(x), size=2, replace=False)
                x_new[idx1], x_new[idx2] = x_new[idx2], x_new[idx1]
                return x_new

            # Generating neighbor solution
            x_new = generate_neighbor_solution(x)

            # Evaluate solution
            if np.all(A @ x_new <= b):
                FO_new = x_new @ c
            else:
                FO_new = -np.inf

            # Update the current solution
            if FO_new > s_new:
                x = x_new
                s_new = FO_new
                best_iteration = n_iterations

                # Check if the current solution is the best found so far
                if s_new > best_value:
                    best_solution = x.copy()
                    best_value = s_new

    # Print the best solution found
    print('The best solution found:')
    print('Objective function value:', round(best_value, 2))

    return


#### 5.	Running the heuristic for the given instance

In [13]:
import time

# Record starting time
start_time = time.time()

# Running the proposed heuristic for this instance
LocalSearch(nVars = coluna, A = A, b = b, c = c, 
            nStarts = 12,
            max_iterations = 35)

# Record ending time
end_time = time.time()

# Calculate the elapsed time
elapsed_time = round(end_time - start_time, 2)

# Print the elapsed time
print("Elapsed time:", elapsed_time, "seconds")


The best solution found:
Objective function value: 180.44
Elapsed time: 123.05 seconds


**Test Run Results**

The best solution found:

- Objective function value: 180.44
- Elapsed time: 123.05 seconds


### End